### PR TITLE
Tickets/dm 39080

### PR DESCRIPTION
**Pull request by iagaponenko**

### CODE REVIEW

File: src/wpublish/QueriesAndChunks.cc [hunk 1]
```diff
@@ -74,15 +79,34 @@ nlohmann::json QueryStatistics::getJsonHist() const {
...
+            bool const selectedByQueryId =
+                    ids.empty() || find(ids.begin(), ids.end(), task->getQueryId()) != ids.end();
+            bool const selectedByTaskState =
+                    states.empty() || find(states.begin(), states.end(), task->state()) != states.end();
+            if (selectedByQueryId && selectedByTaskState) {
+                ++numTasksSelected;
+                // Stop reporting tasks if the limit has been reached. Just keep
+                // counting the number of tasks selected by the filter.
+                if ((taskSelector.maxTasks == 0) || (numTasksSelected < taskSelector.maxTasks)) {
+                    taskEntriesJson.push_back(task->getJson());
```
Comment (jgates108) [inline src/wpublish/QueriesAndChunks.cc:102]: Sending back incomplete lists of tasks could have pretty serious consequences for the displayed dashboard statistics, which could be extremely confusing. There should probably be some indication on the dashboard that it is displaying data for incomplete lists of tasks.

File: src/proto/worker.proto [hunk 2]
```diff
@@ -321,6 +321,30 @@ message WorkerCommandSetChunkListR {
...
+    // Narrow a scope of the request that involves tasks (the above-define parameter
+    // 'include_tasks = true') to a specific collection of queries.
+    repeated uint64 query_ids = 2;
+
+    // An optional selector for the state(s) of tasks. If the collection
+    // is empty then the selector is disregarded. Values to be put into
+    // the collection are the integer representation of the following transient
+    // C++ enum type 'wbase::TaskState'. These values are expected to be
+    // validated and upconverted into the enum class by the worker.
+    repeated uint64 task_states = 3;
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/proto/worker.proto:340]: Yes, exactly. And there are two specific use cases that are needed (used by) the Web Dashboard when selecting tasks by state. The first case is for tasks that are being processed:
```
(EXECUTING_QUERY, READING_DATA)
```
And the next combination is for tasks that haven't been finished:
```
(CREATED, QUEUED, EXECUTING_QUERY,  READING_DATA)
```
I can demo these scenarios at the meeting

File: src/replica/HttpRequestQuery.cc [hunk 3]
```diff
@@ -24,113 +24,140 @@
...
+    throw invalid_argument("HttpRequestQuery::" + func + " mandatory parameter '" + param + "' is missing");
+}
+
+template <typename T>
+T parseRestrictedIntegerType(string const& func, string const& param, string const& str) {
+    int64_t val = stoll(str);
+    if ((val < numeric_limits<T>::min()) || (val > numeric_limits<T>::max())) {
+        throw out_of_range("HttpProcessor::" + func + " value of parameter: " + param +
+                           " exceeds allowed range for the target type");
+    }
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/replica/HttpRequestQuery.cc:48]: Yes, all exceptions thrown by the HTTP modules of the Controller's REST services are caught at the very base class of the modules. Then a JSON object with an explanation of the error is sent back to a client (calling the corresponding REST service). So, a reason for all these exceptions (unless they are caught in the calling context) is to notify clients of the REST services.

In this particular case, the message is telling a caller of the REST service that the service was called with an incorrect value of the parameter.

File: src/replica/HttpRequestQuery.cc [hunk 4]
```diff
@@ -24,113 +24,140 @@
...
 #include <stdexcept>
 
 using namespace std;
 
+namespace {
+
+void throwIf(bool cond, string const& func, string const& param) {
+    if (!cond) return;
+    throw invalid_argument("HttpRequestQuery::" + func + " mandatory parameter '" + param + "' is missing");
+}
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/replica/HttpRequestQuery.cc:40]: Yes, it's a good way to reduce clutter in the code.

File: src/replica/Performance.cc [hunk 5]
```diff
@@ -76,13 +55,13 @@ void Performance::update(ProtocolPerformance const& workerPerformanceInfo) {
...
     uint64_t const t = c_start_time;
-    c_start_time = PerformanceUtils::now();
+    c_start_time = util::TimeUtils::now();
     return t;
 }
 
 uint64_t Performance::setUpdateFinish() {
     uint64_t const t = c_finish_time;
-    c_finish_time = PerformanceUtils::now();
+    c_finish_time = util::TimeUtils::now();
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/replica/Performance.cc:64]: Actually, I didn't move the class `PerformanceUtils`. I only moved one method of that class. Class `PerformanceUtils` is used for performance monitoring of the Replication system's "requests" and "jobs" (async algorithms). Te moved method `PerformanceUtils::now()` didn't really belong here. That's why I moved it to share the method by other modules.

File: src/wpublish/QueriesAndChunks.cc (hunk 1 above)
Comment (iagaponenko) [reply src/wpublish/QueriesAndChunks.cc:102]: Yes, this has already been implemented exactly as you say. I will demo the new version of the Dashboard at the group meeting.

File: src/replica/HttpQservMonitorModule.h [hunk 6]
```diff
@@ -101,6 +106,30 @@ class HttpQservMonitorModule : public HttpModule {
...
+    /**
+     * The helper method for processing the input JSON object and populating
+     * the output collections.
+     * @note The method is shared by implementations of _worker() and _workers(), and
+     * it's needed to avoid code duplication.
+     */
+    void _processWorkerInfo(std::string const& worker, bool success, bool keepResources,
+                            nlohmann::json const& inWorkerInfo, nlohmann::json& statusRef,
+                            std::map<std::string, std::set<int>>& schedulers2chunks,
+                            std::set<int>& chunks) const;
```
[the remark this answers is not in the document]
Comment (iagaponenko) [reply src/replica/HttpQservMonitorModule.h:123]: `success` is an input flag that's passed into the method along with `nlohmann::json const& inWorkerInfo`. The flag is needed to instruct the method on how to process the worker information object. If the flag is set to `false` the method would ignore the worker info by assuming there was a failure to obtain info from the corresponding worker. In theory, I may eliminate that extra flag by passing an empty `json` object into the method to indicate a failure mode.

**Review by fritzm — APPROVED** [review]
LGTM -- thanks!

**Pull request merged**